Source/Utils/ParmParseWithOptionalGroup.cpp

```diff
@@ -0,0 +1,223 @@
+{
+    // First, query name with the group name
+    const amrex::ParmParse pp_with_group(prefix_dot_group);
+    bool is_specified = utils::parser::queryWithParser(pp_with_group, name, ref);
+    if (!is_specified && !group.empty()) {
+        // If it wasn't found, query it with only the prefix
+        const amrex::ParmParse pp_prefix(prefix);
+        is_specified = utils::parser::queryWithParser(pp_prefix, name, ref);
+    }
+    return is_specified;
+}
+
+bool ParmParseWithOptionalGroup::queryWithParser (const char* name, double& ref) const
+{
+    // First, query name with the group name
+    const amrex::ParmParse pp_with_group(prefix_dot_group);
+    bool is_specified = utils::parser::queryWithParser(pp_with_group, name, ref);
+    if (!is_specified && !group.empty()) {
+        // If it wasn't found, query it with only the prefix
+        const amrex::ParmParse pp_prefix(prefix);
+        is_specified = utils::parser::queryWithParser(pp_prefix, name, ref);
+    }
+    return is_specified;
+}
+
+bool ParmParseWithOptionalGroup::queryWithParser (const char* name, float& ref) const
+{
+    // First, query name with the group name
+    const amrex::ParmParse pp_with_group(prefix_dot_group);
+    bool is_specified = utils::parser::queryWithParser(pp_with_group, name, ref);
+    if (!is_specified && !group.empty()) {
+        // If it wasn't found, query it with only the prefix
+        const amrex::ParmParse pp_prefix(prefix);
+        is_specified = utils::parser::queryWithParser(pp_prefix, name, ref);
+    }
+    return is_specified;
+}
+
+bool ParmParseWithOptionalGroup::query (const char* name, std::string& ref) const
+{
+    // First, query name with the group name
+    const amrex::ParmParse pp_with_group(prefix_dot_group);
+    bool is_specified = pp_with_group.query(name, ref);
+    if (!is_specified && !group.empty()) {
+        // If it wasn't found, query it with only the prefix
+        const amrex::ParmParse pp_prefix(prefix);
+        is_specified = pp_prefix.query(name, ref);
+    }
+    return is_specified;
+}
+
+void ParmParseWithOptionalGroup::get_long_string (const char* name, std::string& ref) const
+{
+    // First, query name with the group name
+    const amrex::ParmParse pp_with_group(prefix_dot_group);
+    bool is_specified = pp_with_group.contains(name);
+    if (is_specified) {
+        utils::parser::Store_parserString(pp_with_group, name, ref);
+    } else {
+        // If it wasn't found, query it with only the prefix
+        const amrex::ParmParse pp_prefix(prefix);
+        utils::parser::Store_parserString(pp_prefix, name, ref);
+    }
+
+}
+
+void ParmParseWithOptionalGroup::getWithParser (const char* name, int& ref) const
+{
+    // First, query name with the group name
+    const amrex::ParmParse pp_with_group(prefix_dot_group);
+    bool is_specified = utils::parser::queryWithParser(pp_with_group, name, ref);
+    if (!is_specified) {
```

Comment:
Since this is a `get` the parameter must be supplied. So if `is_speficied` is false and the group is empty, this is an error since the parameter wasn't given. As I have, it then calls `get` which would raise an exception. Though, maybe this could be rearranged so that the exception would reference the parameter name with the group name included - this might be more user friendly.